### PR TITLE
TeraVM Integration LF

### PR DESCRIPTION
**Pull request by vaishaliCI**
Please review new .yaml file, file will perform required steps in LF Jenkins :- 
          1. create Teravm job.
          2. Trigger test run when magma code will push in GitHub.
          3. Send notifications in slack for test start, fail, abort, success.
          4. Send logs on email ID.

@fabiopalumbo

### CODE REVIEW

File: jjb/ci-teravm/ci-teravm.yaml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            Jenkins Build URL:
+            $BUILD_URL
+          unstable: true
+          fixed: true
+          always: true
+          attachments: "**/reports/*"
+          send-to:
+            - recipients
+      - slack:
+          team-domain: 'magmacore'
```
Comment (fabiopalumbo) [inline jjb/ci-teravm/ci-teravm.yaml:45]: Change this single quote to double quote

File: jjb/ci-teravm/ci-teravm.yaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+          auth-token: 'DPKFwmiCxQ9VPq95liGUCEwk'
```
Comment (fabiopalumbo) [inline jjb/ci-teravm/ci-teravm.yaml:46]: Change this single quote to double quote

File: jjb/ci-teravm/ci-teravm.yaml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+          auth-token-id: 'magma-test'
```
Comment (fabiopalumbo) [inline jjb/ci-teravm/ci-teravm.yaml:47]: Change this single quote to double quote

File: jjb/ci-teravm/ci-teravm.yaml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+          notify-failure: True
+          #notify-every-failure: True
+          #notify-back-to-normal: True
+          #notify-repeated-failure: True
+          #notify-regression: True
+          #include-test-summary: True
+          #include-failed-tests: True
+          commit-info-choice: "AUTHORS_AND_TITLES"
+          include-custom-message: True
+          custom-message: 'A teravm integration custom message.'
```
Comment (fabiopalumbo) [inline jjb/ci-teravm/ci-teravm.yaml:64]: Change this single quote to double quote

**Comment by fabiopalumbo** [comment]
Remember that you can run the precommit check of tox locally with

`pre-commit run --all-files`

**Comment by fpalumbo-phinx** [comment]
recheck

**Comment by fpalumbo-phinx** [comment]
jjb-deploy

**Comment by fabiopalumbo** [comment]
jjb-deploy

**Comment by fabiopalumbo** [comment]
jjb-deploy LF-teravm

**Review by fabiopalumbo — DISMISSED** [review]

**Comment by fabiopalumbo** [comment]
So we created the job template, now we need to create the yaml file that will call the template

Changes to be made
* change folder name from ci-teravm change to teravm
* rename the ci-teravm.yml to teravm-templates.yaml and move it to a subfolder called templates
* change the project name in teravm-templates.yaml to add a variable, instead of LF-teravm -->  {project-name}-LF-teravm
* Create a new teravm.yaml to call the job template

**Comment by fabiopalumbo** [comment]
reverify

**Comment by fabiopalumbo** [comment]
jjb-deploy teravm-LF-teravm

File: jjb/teravm/teravm.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+- project:
+    name: teravm-jobs
+    jobs:
+      - "{project-name}-LF-teravm"
+    project: teravm
+    project-name: teravm
+    archive-artifacts: "**/*.log"
+    branch: main
+    build-timeout: 60
+    build-node: magma-driver
```
Comment (fabiopalumbo) [inline jjb/teravm/teravm.yaml:15]: here should be the name of the node  `static-ubuntu-dev-teravm`

**Comment by fabiopalumbo** [comment]
recheck

**Comment by fabiopalumbo** [comment]
jjb-deploy teravm-LF-teravm

**Review by fabiopalumbo — DISMISSED** [review]

**Comment by fabiopalumbo** [comment]
reverify

**Comment by fabiopalumbo** [comment]
jjb-deploy teravm-LF-teravm

**Comment by fabiopalumbo** [comment]
```
INFO:root:Caching type triggers of triggers = jenkins_jobs.modules.triggers:Triggers
INFO:root:Caching type builders of builders = jenkins_jobs.modules.builders:Builders
INFO:root:Caching type wrappers of wrappers = jenkins_jobs.modules.wrappers:Wrappers
INFO:jenkins_jobs.builder:Number of jobs generated:  1
INFO:jenkins_jobs.builder:Creating jenkins job teravm-LF-teravm
```

File: jjb/teravm/teravm.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+- project:
+    name: teravm-jobs
+    jobs:
+      - "{project-name}-LF-teravm"
+    project: teravm
+    project-name: teravm
+    archive-artifacts: "**/*.log"
+    branch: main
+    build-timeout: 60
+    build-node: static-ubuntu-dev-teravm
```
Comment (fabiopalumbo) [inline jjb/teravm/teravm.yaml:15]: You need to change here the build-node for `ci-teravm`

**Review by fabiopalumbo — DISMISSED** [review]

**Comment by fpalumbo-phinx** [comment]
The error still shows
`02:05:37 ERROR: unknown build-node static-ubuntu-dev-teravm in jjb/teravm/teravm.yaml
02:05:37 INFO: 3 valid label(s), 1 invalid label(s)`

**Comment by fabiopalumbo** [comment]
reverify

**Comment by jwagantall** [comment]
reverify

**Comment by jwagantall** [comment]
reverify

**Comment by fabiopalumbo** [comment]
reverify

**Comment by fabiopalumbo** [comment]
reverify

**Review by fabiopalumbo — DISMISSED** [review]

File: jjb/teravm/templates/teravm-templates.yaml [hunk 7]
```diff
@@ -0,0 +1,67 @@
+- job-template:
+    name: "{project-name}-LF-teravm"
+    node: "ci-teravm"
+    build-args: ""
+
```
Comment (fabiopalumbo) [inline jjb/teravm/templates/teravm-templates.yaml:5]: We need to add now in the job template the publisher

```
- job-template:
    name: '{name}-unit-tests'
    builders:
    - shell: unittest
    publishers:
    - email:
        recipients: '{mail-to}'
```

Ref. https://docs.openstack.org/infra/jenkins-job-builder/publishers.html#publishers.slack

**Review by fpalumbo-phinx — DISMISSED** [review]

**Review by fabiopalumbo — DISMISSED** [review]

**Review by fabiopalumbo — DISMISSED** [review]

**Comment by fpalumbo-celsius** [comment]
re-verify

**Comment by fabiopalumbo** [comment]
Example of use of generic webhook trigger

 ```
   triggers:
      - generic-webhook-trigger:
            post-content-params:
                - type: JSONPath
                  key: action
                  value: $.action
                - type: JSONPath
                  key: merged
                  value: $.pull_request.merged
                - type: JSONPath
                  key: repoUrl
                  value: $.pull_request.base.repo.html_url
                - type: JSONPath
                  key: repoName
                  value: $.pull_request.base.repo.name
                - type: JSONPath
                  key: branchName
                  value: $.pull_request.base.ref
                - type: JSONPath
                  key: commitHash
                  value: $.pull_request.merge_commit_sha
            regex-filter-text: $action,$merged
            regex-filter-expression: ^(closed,true)$
            cause: Generic Cause
            token: '{project}'
```

**Pull request closed without merging**